File: application/api/datamodel/entity-instance-change-event.js
/* global Java */

define(['boxing'], function(B) {
    var className = "com.eas.client.model.application.EntityInstanceChangeEvent";
    var javaClass = Java.type(className);
    /**
     * Generated constructor.
     * @constructor EntityInstanceChangeEvent EntityInstanceChangeEvent
     */
    function EntityInstanceChangeEvent() {
        var maxArgs = 0;
        var delegate = arguments.length > maxArgs ?
              arguments[maxArgs] 
            : new javaClass();

        Object.defineProperty(this, "unwrap", {
            configurable: true,
            value: function() {
                return delegate;
            }
        });
        if(EntityInstanceChangeEvent.superclass)
            EntityInstanceChangeEvent.superclass.constructor.apply(this, arguments);
        delegate.setPublished(this);
        /**
         * The new value.
         */
        this.newValue = new Object();
        Object.defineProperty(this, "newValue", {
            get: function() {
                var value = delegate.newValue;
                return B.boxAsJs(value);
            }
        });

        /**
         * The changed property name.
         */
        this.propertyName = '';
        Object.defineProperty(this, "propertyName", {
            get: function() {
                var value = delegate.propertyName;
                return B.boxAsJs(value);
            }
        });

        /**
         * The old value.
         */
        this.oldValue = new Object();
        Object.defineProperty(this, "oldValue", {
            get: function() {
                var value = delegate.oldValue;
                return B.boxAsJs(value);
            }
        });

        /**
         * The source object of the event.
         */
        this.source = new Object();
        Object.defineProperty(this, "source", {
            get: function() {
                var value = delegate.source;
                return B.boxAsJs(value);
            }
        });

        /**
         * The updated element.
         */
        this.object = new Object();
        Object.defineProperty(this, "object", {
            get: function() {
                var value = delegate.object;
                return B.boxAsJs(value);
            }
        });

    }

    var ScriptsClass = Java.type("com.eas.script.Scripts");
    var space = ScriptsClass.getSpace();
    space.putPublisher(className, function(aDelegate) {
        return new EntityInstanceChangeEvent(aDelegate);
    });
    return EntityInstanceChangeEvent;
});

File: data-model/src/com/eas/client/model/application/EntityInstanceChangeEvent.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.eas.client.model.application;

import com.eas.client.events.PublishedSourcedEvent;
import com.eas.client.metadata.Field;
import com.eas.script.HasPublished;
import com.eas.script.NoPublisherException;
import com.eas.script.ScriptFunction;
import com.eas.script.Scripts;
import jdk.nashorn.api.scripting.JSObject;

/**
 * Represents an entity's instance change event.
 *
 * @author vv
 */
public class EntityInstanceChangeEvent extends PublishedSourcedEvent {

    protected Field field;
    protected Object oldValue;
    protected Object newValue;

    public EntityInstanceChangeEvent(HasPublished aSource, Field aField, Object aOldValue, Object aNewValue) {
        super(aSource);
        field = aField;
        oldValue = aOldValue;
        newValue = aNewValue;
    }

    public Field getField() {
        return field;
    }

    private static final String PROPERTY_NAME_JSDOC = ""
            + "/**\n"
            + " * The changed property name.\n"
            + " */";

    @ScriptFunction(jsDoc = PROPERTY_NAME_JSDOC)
    public String getPropertyName() {
        return field != null ? field.getName() : null;
    }

    private static final String OLD_VALUE_JSDOC = ""
            + "/**\n"
            + " * The old value.\n"
            + " */";

    @ScriptFunction(jsDoc = OLD_VALUE_JSDOC)
    public Object getOldValue() {
        return oldValue;
    }

    private static final String NEW_VALUE_JSDOC = ""
            + "/**\n"
            + " * The new value.\n"
            + " */";

    @ScriptFunction(jsDoc = NEW_VALUE_JSDOC)
    public Object getNewValue() {
        return newValue;
    }

    private static final String OBJECT_JSDOC = ""
            + "/**\n"
            + " * The updated element.\n"
            + " */";

    @ScriptFunction(jsDoc = OBJECT_JSDOC)
    public HasPublished getObject() {
        return source;
    }

    @Override
    public JSObject getPublished() {
        if (published == null) {
            JSObject publisher = Scripts.getSpace().getPublisher(this.getClass().getName());
            if (publisher == null || !publisher.isFunction()) {
                throw new NoPublisherException();
            }
            published = (JSObject) publisher.call(null, new Object[]{this});
        }
        return published;
    }

}
